src/phasorpy/phasor.py

```diff
@@ -34,23 +35,216 @@
+        `sample_phase` size does not equal the number of samples along `axis`.
+    TypeError
+        The `signal` or `dtype` types are not supported.
+
+    Examples
+    --------
+    Calculate phasor coordinates of a phase-shifted sinusoidal waveform:
+
+    >>> sample_phase = numpy.linspace(0, 2 * math.pi, 5, endpoint=False)[::-1]
+    >>> signal = 1.1 * (
+    ...     numpy.cos(sample_phase - 0.78539816) * 2 *  0.70710678 + 1
+    ... )
+    >>> phasor_from_signal_f1(
+    ...    signal, sample_phase=sample_phase
+    ... )  # doctest: +NUMBER
+    (1.1, 0.5, 0.5)
+
+    """
+    signal = numpy.array(signal, order='C', ndmin=1, copy=False)
+    # assert signal.flags['C_CONTIGUOUS']
+    samples = signal.shape[axis]  # this also verifies axis
+    if samples < 3:
+        raise ValueError(f'not enough {samples=} along {axis=}')
+    if sample_phase is None:
+        sample_phase = numpy.linspace(
+            0, 2 * math.pi, samples, endpoint=False, dtype=numpy.float64
+        )
+    else:
+        sample_phase = numpy.array(
+            sample_phase, dtype=numpy.float64, copy=False, ndmin=1
+        )
+        if sample_phase.ndim != 1 or sample_phase.size != samples:
+            raise ValueError(f'{sample_phase.shape=} != ({samples},)')
+    if num_threads is None:
+        num_threads = 1
+    elif num_threads == 0:
+        try:
+            num_threads = len(os.sched_getaffinity(0))  # type: ignore
+        except AttributeError:
+            # sched_getaffinity not available on Windows
+            num_threads = os.cpu_count()
+            if num_threads is None:
+                num_threads = 1
+        num_threads = max(num_threads // 2, 1)
+    elif num_threads < 0:
+        raise ValueError(f'{num_threads=} < 0')
+
+    # pure numpy implementation for reference:
+    # shape = [1] * signal.ndim
+    # shape[axis] = sample_phase.size
+    # sample_phase = sample_phase.reshape(shape)  # make broadcastable
+    # mean = numpy.mean(signal, axis=axis)
+    # real = numpy.mean(signal * numpy.cos(sample_phase), axis=axis)
+    # real /= mean
+    # imag = numpy.mean(signal * numpy.sin(sample_phase), axis=axis)
+    # imag /= mean
+    # return mean, real, imag
+
+    sincos = numpy.empty((samples, 2))
+    sincos[:, 0] = numpy.cos(sample_phase)
+    sincos[:, 1] = numpy.sin(sample_phase)
+    # reshape to 3D with axis in middle
+    axis = axis % signal.ndim
+    shape0 = signal.shape[:axis]
+    shape1 = signal.shape[axis + 1 :]
+    size0 = math.prod(shape0)
+    size1 = math.prod(shape1)
+    phasor = numpy.empty((3, size0, size1), dtype)
+    signal = signal.reshape((size0, samples, size1))
+
+    _phasor_from_signal(phasor, signal, sincos, num_threads)
+
+    # restore original shape
+    shape = shape0 + shape1
+    mean = phasor[0].reshape(shape)
+    real = phasor[1].reshape(shape)
+    imag = phasor[2].reshape(shape)
+    if shape:
+        return mean, real, imag
+    return mean.item(), real.item(), real.item()
```

Comment:
In this case the real component is returned as if it was the imaginary?

Reply:
Ouch. That should have been caught by the tests.

Reply:
Thanks again for catching this. It is fixed and tested now.